src/cli/cmd/project/clone.rs

```diff
@@ -1,22 +1,207 @@
+                .await
+                .context(HttpClientSnafu)?,
+        };
+        if let Some(details) = opt_details {
+            let target = self.target_dir()?.join(&details.slug);
+            let renku_project_cfg = RenkuProjectConfig::new(
+                &ctx.renku_url,
+                ProjectInfo {
+                    id: details.id.clone(),
+                    namespace: details.namespace.clone(),
+                    slug: details.slug.clone(),
+                },
+            );
+
+            ctx.write_err(&SimpleMessage {
+                message: format!(
+                    "Cloning {} ({}) into {}...",
+                    details.slug,
+                    details.id,
+                    &target.display()
+                ),
+            })
+            .await
+            .context(WriteResultSnafu)?;
+
+            write_config(renku_project_cfg, &target).await?;
+
+            let ctx = clone_project(ctx, &details, target).await?;
+            ctx.write_result(&details).await.context(WriteResultSnafu)?;
+        } else {
+            ctx.write_err(&SimpleMessage {
+                message: format!("Project '{}' doesn't exist.", &project_id),
+            })
+            .await
+            .context(WriteResultSnafu)?;
+        }
         Ok(())
     }
+
+    fn project_id(&self) -> Result<ProjectId, Error> {
+        self.project_ref
+            .parse::<ProjectId>()
+            .context(ProjectIdParseSnafu)
+    }
+
+    fn target_dir(&self) -> Result<PathBuf, Error> {
+        match &self.target_dir {
+            Some(dir) => Ok(std::path::PathBuf::from(dir)),
+            None => std::env::current_dir().context(CurrentDirSnafu),
+        }
+    }
 }
 
-#[derive(Debug, Snafu)]
-pub enum Error {
-    Dummy,
+async fn clone_project<'a>(
+    ctx: Context,
+    project: &ProjectDetails,
+    target: PathBuf,
+) -> Result<Context, Error> {
+    tokio::fs::create_dir_all(&target)
+        .await
+        .context(CreateDirSnafu)?;
+
+    let mut tasks = JoinSet::new();
+    let cc = Arc::new(ctx);
+    let tt = Arc::new(target);
+    for repo in project.repositories.iter() {
+        let cc = cc.clone();
+        let tt = tt.clone();
+        let rr = repo.to_string();
+        tasks.spawn(clone_repository(cc, rr, tt));
+    }
+
+    while let Some(res) = tasks.join_next().await {
+        res.context(TaskJoinSnafu)??;
+    }
+    Ok(Arc::into_inner(cc).unwrap())
+}
+
+async fn clone_repository(
+    ctx: Arc<Context>,
+    repo_url: String,
+    dir: Arc<PathBuf>,
+) -> Result<(), Error> {
+    let name = match repo_url.rsplit_once('/') {
+        Some((_, n)) => n,
+        None => "no-name",
+    };
+    let local_path = dir.join(name);
+    if local_path.exists() {
+        ctx.write_err(&SimpleMessage {
+            message: format!("The repository {} already exists", name),
+        })
+        .await
+        .context(WriteResultSnafu)?;
+    } else {
```

Comment:
issue/question: this logic is a bit odd. if there were multiple repos without a name, it would clone one into `/no-name` and then print `The repository 'no-name' already exists` for subsequent ones. That seems pretty confusing from a users perspective.
There'd be no way for a user to know what happened, which repo caused it and what they could do about it.
I wonder if we should just error out here? or have more logic to get a reasonable name from a giturl if it fails the rsplit?

Reply:
yes good catch, thank you! I think we should error out immediately - the name should be always there. And in case the user only forgot to type it, it would be nicer to show the error right away, wdyt?